### PR TITLE
Add gesture handling

### PR DESCRIPTION
**Pull request by CendioOssman**
Handle a bunch of different gestures in order to easily support several mouse actions (clicking different buttons, scroll wheel).

Also aiming to handle gestures locally for panning and zooming.

### CODE REVIEW

File: core/input/gesturehandler.js [hunk 1]
```diff
@@ -338,31 +358,76 @@ export default class GestureHandler {
...
+    }
+
+    _stopLongpressTimeout() {
+        clearTimeout(this._longpressTimeoutId);
+        this._longpressTimeoutId = null;
+    }
+
+    _longpressTimeout() {
+        if (this._hasDetectedGesture()) {
+            Log.Error("A longpress gesture failed, conflict with a different gesture");
```
Comment (CendioOssman) [inline core/input/gesturehandler.js:374]: Is this equivalent to an assert? Perhaps we should throw an error in such cases?

File: core/input/gesturehandler.js [hunk 2]
```diff
@@ -338,31 +358,76 @@ export default class GestureHandler {
...
+        this._twoTouchTimeoutId = null;
+    }
+
+    _isTwoTouchTimeoutRunning() {
+        return this._twoTouchTimeoutId !== null;
+    }
+
+    _twoTouchTimeout() {
+        if (this._tracked.length === 0) {
+            Log.Error("A pinch or two drag gesture failed, no tracked touches");
```
Comment (CendioOssman) [inline core/input/gesturehandler.js:399]: Dito.

File: tests/test.gesturehandler.js [hunk 3]
```diff
@@ -34,43 +34,50 @@ describe('Gesture handler', function () {
...
+        touch.clientY = y;
         let ev = { type: 'touchmove',
-                                  touches: touches,
-                                   targetTouches: touches,
-                                   changedTouches: [ touch ] };
+                   touches: touches,
+                   targetTouches: touches,
+                   changedTouches: [ touch ] };
+        ev.stopPropagation = sinon.spy();
+        ev.preventDefault = sinon.spy();
```
Comment (CendioOssman) [inline tests/test.gesturehandler.js:63]: Nitpick, but why not include these in the object literal?

File: core/input/gesturehandler.js (hunk 1 above)
Comment (samhed) [reply core/input/gesturehandler.js:374]: Yes, that's a good idea

File: core/input/gesturehandler.js [hunk 4]
```diff
@@ -0,0 +1,562 @@
...
+            if (this._tracked[i].active) {
+                this._ignored.push(this._tracked[i].id);
+            }
+        }
+        this._tracked = [];
+
+        this._state = GH_NOGESTURE;
+
+        if (this._ignored.indexOf(id) !== -1) {
+            this._ignored.splice(this._ignored.indexOf(id), 1);
```
Comment (samhed) [inline core/input/gesturehandler.js:338]: Perhaps this needs a comment explaining what it does

File: core/rfb.js [hunk 5]
```diff
@@ -11,12 +11,14 @@ import { toUnsigned32bit, toSigned32bit } from './util/int.js';
 import * as Log from './util/logging.js';
 import { encodeUTF8, decodeUTF8 } from './util/strings.js';
 import { dragThreshold } from './util/browser.js';
+import { clientToElement } from './util/element.js';
```
Comment (samhed) [inline core/rfb.js:14]: To me this seems to fit in `util/browser.js`, what's the reasoning here?

File: core/rfb.js [hunk 6]
```diff
@@ -36,6 +38,12 @@ import TightPNGDecoder from "./decoders/tightpng.js";
 const DISCONNECT_TIMEOUT = 3;
 const DEFAULT_BACKGROUND = 'rgb(40, 40, 40)';
 
+// Gesture thresholds
+const ZOOMSENS = 30;
+const SCRLSENS = 50;
```
Comment (samhed) [inline core/rfb.js:43]: These two should get some sort of prefix, like `GESTURE_ZOOMSENS`

File: core/rfb.js [hunk 7]
```diff
@@ -67,6 +75,12 @@ export default class RFB extends EventTargetMixin {
         this._target = target;
         this._url = url;
 
+        this._gestureLastTapTime = null;
+        this._gestureFirstDoubleTapEvent = null;
```
Comment (samhed) [inline core/rfb.js:79]: That's a very long name, I'd say `_gestureFirstDoubleTap` is enough.

File: core/rfb.js [hunk 8]
```diff
@@ -876,6 +904,156 @@ export default class RFB extends EventTargetMixin {
         RFB.messages.pointerEvent(this._sock, this._display.absX(x), this._display.absY(y), this._mouse_buttonMask);
     }
 
+    _handleTapEvents(ev, bmask) {
```
Comment (samhed) [inline core/rfb.js:907]: Should be `_handleTapEvent`, not plural

File: tests/test.rfb.js [hunk 9]
```diff
@@ -105,9 +106,12 @@ describe('Remote Frame Buffer Protocol Client', function () {
             this._rQ = rQ;
         };
 
+        // Avoiding printing the entire Websock buffer on errors
+        Websock.prototype.toString = function () { return "[object Websock]"; };
```
Comment (samhed) [inline tests/test.rfb.js:110]: Awesome! Thanks

File: tests/test.rfb.js [hunk 10]
```diff
@@ -2864,6 +2868,637 @@ describe('Remote Frame Buffer Protocol Client', function () {
...
+                let bounds = client._canvas.getBoundingClientRect();
+
+                res.x = bounds.left + x;
+                res.y = bounds.top + y;
+
+                return res;
+            }
+
+            function gestureStart(gestureType, x, y,
+                magnitudeX = 0, magnitudeY = 0) {
```
Comment (samhed) [inline tests/test.rfb.js:2897]: bad indentation

File: tests/test.rfb.js [hunk 11]
```diff
@@ -2864,6 +2868,637 @@ describe('Remote Frame Buffer Protocol Client', function () {
...
+
+                detail.magnitudeX = magnitudeX;
+                detail.magnitudeY = magnitudeY;
+
+                let ev = new CustomEvent('gesturestart', { detail: detail });
+                client._canvas.dispatchEvent(ev);
+            }
+
+            function gestureMove(gestureType, x, y,
+                magnitudeX = 0, magnitudeY = 0) {
```
Comment (samhed) [inline tests/test.rfb.js:2909]: indentation

File: tests/test.rfb.js [hunk 12]
```diff
@@ -2864,6 +2868,637 @@ describe('Remote Frame Buffer Protocol Client', function () {
...
+                                                                           20, 40, 0x0);
+                    expect(pointerEvent.secondCall).to.have.been.calledWith(client._sock,
+                                                                            20, 40, bmask);
+
+                    pointerEvent.resetHistory();
+
+                    gestureMove('drag', 30, 50);
+
+                    expect(pointerEvent).to.have.been.calledOnce;
+
```
Comment (samhed) [inline tests/test.rfb.js:3100]: inconsistent newline

File: tests/test.rfb.js [hunk 13]
```diff
@@ -2864,6 +2868,637 @@ describe('Remote Frame Buffer Protocol Client', function () {
...
+            });
+
+            describe('Gesture long press', function () {
+                it('should handle long press events', function () {
+                    let bmask = 0x4;
+
+                    gestureStart('longpress', 20, 40);
+
+                    expect(pointerEvent).to.have.been.calledTwice;
+
```
Comment (samhed) [inline tests/test.rfb.js:3123]: remove newline

File: tests/test.rfb.js [hunk 14]
```diff
@@ -2864,6 +2868,637 @@ describe('Remote Frame Buffer Protocol Client', function () {
...
+                                                                            20, 40, bmaskScrlUp);
+                    expect(pointerEvent.getCall(2)).to.have.been.calledWith(client._sock,
+                                                                            20, 40, 0x0);
+                    expect(pointerEvent.getCall(3)).to.have.been.calledWith(client._sock,
+                                                                            20, 40, bmaskScrlRight);
+                    expect(pointerEvent.getCall(4)).to.have.been.calledWith(client._sock,
+                                                                            20, 40, 0x0);
+                });
+
+                it('should handle multiple small gesture twodrag events', function () {
```
Comment (samhed) [inline tests/test.rfb.js:3258]: Should we have a test for a large pan movement as well? We have a `while` loop that sends multiple scrolls if the delta is large enough afaik

File: tests/test.rfb.js [hunk 15]
```diff
@@ -2864,6 +2868,637 @@ describe('Remote Frame Buffer Protocol Client', function () {
...
+                    expect(pointerEvent).to.have.been.calledThrice;
+                    expect(pointerEvent.firstCall).to.have.been.calledWith(client._sock,
+                                                                           20, 40, 0x0);
+                    expect(pointerEvent.secondCall).to.have.been.calledWith(client._sock,
+                                                                            20, 40, bmask);
+                    expect(pointerEvent.thirdCall).to.have.been.calledWith(client._sock,
+                                                                           20, 40, 0x0);
+                });
+
+                it('should not keep same position for multiple onetap events if too far away', function () {
```
Comment (samhed) [inline tests/test.rfb.js:2984]: A similar test is needed for taps with enough time inbetween

File: core/rfb.js (hunk 5 above)
Comment (CendioOssman) [reply core/rfb.js:14]: `browser.js` is about browser capabilities

File: core/rfb.js (hunk 6 above)
Comment (CendioOssman) [reply core/rfb.js:43]: And do we use all-caps on other constants?

File: core/rfb.js (hunk 7 above)
Comment (CendioOssman) [reply core/rfb.js:79]: Then you lose the context of what type of object is stored in this variable. Besides, JavaScript loves long names. :)

File: core/rfb.js (hunk 6 above)
Comment (samhed) [reply core/rfb.js]: It seems like it

fixed in 6d57ee9

File: core/rfb.js (hunk 8 above)
Comment (samhed) [reply core/rfb.js]: fixed in 6d57ee9

File: core/input/gesturehandler.js (hunk 4 above)
Comment (samhed) [reply core/input/gesturehandler.js:342]: added some comments in fc0e11b27823c196ad913a917342340a863f5c12

File: tests/test.rfb.js (hunk 10 above)
Comment (samhed) [reply tests/test.rfb.js]: fixed in f64b89e563978715c840c97be82465ae9c271d56

File: tests/test.rfb.js (hunk 11 above)
Comment (samhed) [reply tests/test.rfb.js]: fixed in f64b89e563978715c840c97be82465ae9c271d56

File: tests/test.rfb.js (hunk 12 above)
Comment (samhed) [reply tests/test.rfb.js]: fixed in f64b89e563978715c840c97be82465ae9c271d56

File: tests/test.rfb.js (hunk 13 above)
Comment (samhed) [reply tests/test.rfb.js]: fixed in f64b89e563978715c840c97be82465ae9c271d56

File: tests/test.rfb.js (hunk 14 above)
Comment (samhed) [reply tests/test.rfb.js:3287]: added in cf4a1988fc0a553bee061ef36563c879a25c3434

File: tests/test.rfb.js (hunk 15 above)
Comment (samhed) [reply tests/test.rfb.js]: added in de79cfc713e463a0f9578dbc7222c4a7f8e81006

File: core/rfb.js (hunk 7 above)
Comment (samhed) [reply core/rfb.js]: Fine, I had already renamed it in 6d57ee9, but it's now fixed in e59e62b

File: tests/test.rfb.js (hunk 10 above)
Comment (samhed) [reply tests/test.rfb.js]: and lint rule fixed in 4395eb3

File: core/input/gesturehandler.js [hunk 16]
```diff
@@ -0,0 +1,567 @@
...
+const GH_THREETAP  = 4;
+const GH_DRAG      = 8;
+const GH_LONGPRESS = 16;
+const GH_TWODRAG   = 32;
+const GH_PINCH     = 64;
+
+const GH_INITSTATE = 127;
+
+const GH_MOVE_THRESHOLD = 50;
+const GH_ANGLE_THRESHOLD = 50; // Degrees
```
Comment (samhed) [inline core/input/gesturehandler.js:22]: This should be 90 degrees

**Comment by CendioNiko** [comment]
Tested this on Windows 10 in the following browsers:
Internet Explorer 11, Firefox 77, Chrome 83 and Edge 44.

Tested that our newly implemented touch gestures works as intended and also that mouse wheel/buttons still works. I also tested that the panning option from the sidebar functioned in combination with touch.

**Comment by CendioAlex** [comment]
Tested touch gestures for iOS on Safari, ChromeOS on Chrome 83 and Linux (running Wayland) on Firefox 77 and Chrome 83. The following works fine:
 - onetap
 - twotap
 - threetap
 - longpress
 - pinch
 - twodrag
 - double tap
 - start with cursor outside and interact with the session
 - iignore new touches during a gesture
 - ViewOnly
 - panning

Also tested that the mouse button/wheel works as before, for every mentioned device and browser

**Comment by juanjoDiaz** [comment]
Was there any reason to move the Mouse handling stuff into RFB instead of keeping it into a separate `input/MouseHandler` class that can be attached/detached like the Keyboard or the Gesture handlers?

Having separate input handlers seems like a much cleaner and maintainable structure to me.

**Comment by CendioOssman** [comment]
There wasn't much of value left in `Mouse` so the abstraction was just making things harder to read.

We can't separate the input fully from `RFB` as the input sometimes goes in to the session, sometimes is used to control local things (e.g. panning). Right now there wasn't really anything around mouse events independent enough to separate out. If something shows up in the future we can always rearrange things then.

**Comment by sportegioco** [comment]
I want to sincerely thank all of you for the addition of the guestures that have allowed the use of novnc in all touch devices much easier.
It would be useful to have a guesture that activates the keyboard view, or even through an icon that appears near the cursor as in the image that I am attaching.
Thanks
Gianpaolo

guesture

**Pull request merged**